go/vt/vtorc/discovery/queue.go

```diff
@@ -33,85 +33,76 @@ import (
 	"vitess.io/vitess/go/vt/vtorc/config"
 )
 
-// Queue contains information for managing discovery requests
-type Queue struct {
-	sync.Mutex
+// queueItem represents an item in the discovery.Queue.
+type queueItem struct {
+	PushedAt time.Time
+	Key      string
+}
 
-	name         string
-	done         chan struct{}
-	queue        chan string
-	queuedKeys   map[string]time.Time
-	consumedKeys map[string]time.Time
+// Queue is an implementation of discovery.Queue.
+type Queue struct {
+	mu       sync.Mutex
+	enqueued map[string]struct{}
+	queue    chan queueItem
 }
 
-// CreateQueue allows for creation of a new discovery queue
-func CreateQueue(name string) *Queue {
+// NewQueue creates a new queue.
+func NewQueue() *Queue {
 	return &Queue{
-		name:         name,
-		queuedKeys:   make(map[string]time.Time),
-		consumedKeys: make(map[string]time.Time),
-		queue:        make(chan string, config.DiscoveryQueueCapacity),
+		enqueued: make(map[string]struct{}),
+		queue:    make(chan queueItem, config.DiscoveryQueueCapacity),
 	}
 }
 
-// QueueLen returns the length of the queue (channel size + queued size)
-func (q *Queue) QueueLen() int {
-	q.Lock()
-	defer q.Unlock()
+// setKeyCheckEnqueued returns true if a key is already enqueued, if
+// not the key will be marked as enqueued and false is returned.
+func (q *Queue) setKeyCheckEnqueued(key string) (alreadyEnqueued bool) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
 
-	return len(q.queue) + len(q.queuedKeys)
+	if _, found := q.enqueued[key]; found {
+		alreadyEnqueued = true
+	} else {
+		q.enqueued[key] = struct{}{}
+	}
+	return alreadyEnqueued
+}
+
+// QueueLen returns the length of the queue.
+func (q *Queue) QueueLen() int {
+	return len(q.queue)
```

Comment:
The previous code did `return len(q.queue) + len(q.queuedKeys)` and I'm not sure why 🤔 

If we see a reason I can restore this but it adds a lock. I think only metrics call this lock